File: packages/sdk-python/veto/integrations/langchain/callback.py
"""
LangChain callback handler for Veto observability.

Provides a ``BaseCallbackHandler`` subclass that logs tool start, end,
and error events. Callbacks are observational and cannot block execution.
Use alongside middleware for audit logging, metrics, or tracing.

Usage::

    from veto.integrations.langchain import VetoCallbackHandler

    handler = VetoCallbackHandler(
        on_tool_start=lambda name, inp: print(f"Started: {name}"),
        on_tool_end=lambda output: print(f"Done: {output}"),
    )

    result = await agent.invoke(
        {"messages": [{"role": "user", "content": "Hello"}]},
        config={"callbacks": [handler]},
    )
"""

from __future__ import annotations

import logging
from typing import Any, Callable, Optional
from uuid import UUID

logger = logging.getLogger("veto.integrations.langchain")


class VetoCallbackHandler:
    """
    Observational callback handler for LangChain tool events.

    Compatible with LangChain's callback system. Fires on tool
    start/end/error for logging and metrics. Does not block execution.

    Args:
        on_tool_start: ``(tool_name: str, input_str: str) -> None``
        on_tool_end: ``(output: str) -> None``
        on_tool_error: ``(error: BaseException) -> None``
    """

    name = "VetoCallbackHandler"

    def __init__(
        self,
        *,
        on_tool_start: Optional[Callable[..., Any]] = None,
        on_tool_end: Optional[Callable[..., Any]] = None,
        on_tool_error: Optional[Callable[..., Any]] = None,
    ) -> None:
        self._on_tool_start = on_tool_start
        self._on_tool_end = on_tool_end
        self._on_tool_error = on_tool_error

    def on_tool_start(
        self,
        serialized: dict[str, Any],
        input_str: str,
        *,
        run_id: UUID,
        parent_run_id: Optional[UUID] = None,
        **kwargs: Any,
    ) -> None:
        name = serialized.get("name", "unknown")
        logger.debug("Tool started: %s", name)
        if self._on_tool_start is not None:
            self._on_tool_start(name, input_str)

    def on_tool_end(
        self,
        output: str,
        *,
        run_id: UUID,
        parent_run_id: Optional[UUID] = None,
        **kwargs: Any,
    ) -> None:
        logger.debug("Tool finished")
        if self._on_tool_end is not None:
            self._on_tool_end(output)

    def on_tool_error(
        self,
        error: BaseException,
        *,
        run_id: UUID,
        parent_run_id: Optional[UUID] = None,
        **kwargs: Any,
    ) -> None:
        logger.debug("Tool error: %s", error)
        if self._on_tool_error is not None:
            self._on_tool_error(error)

File: packages/sdk/src/integrations/langchain/callback.ts
/**
 * LangChain callback handler for Veto logging and observability.
 *
 * Callbacks are observational — they fire after tool start/end/error
 * and cannot block execution. Use this alongside middleware for
 * audit logging, metrics, or tracing.
 *
 * @example
 * ```ts
 * import { createVetoCallbackHandler } from 'veto-sdk/integrations/langchain';
 *
 * const handler = createVetoCallbackHandler({
 *   onToolStart: (name, input) => console.log(`Tool started: ${name}`),
 *   onToolEnd: (name, output) => console.log(`Tool finished: ${name}`),
 *   onToolError: (name, error) => console.error(`Tool failed: ${name}`, error),
 * });
 *
 * const result = await agent.invoke(
 *   { messages: [{ role: 'user', content: 'Hello' }] },
 *   { callbacks: [handler] },
 * );
 * ```
 */
export interface VetoCallbackOptions {
  onToolStart?: (toolName: string, input: string) => void | Promise<void>;
  onToolEnd?: (toolName: string, output: string) => void | Promise<void>;
  onToolError?: (toolName: string, error: Error) => void | Promise<void>;
}

export function createVetoCallbackHandler(options: VetoCallbackOptions): Record<string, any> {
  const toolNames = new Map<string, string>();

  return {
    name: 'VetoCallbackHandler',

    async handleToolStart(
      tool: { id?: string[]; name?: string },
      input: string,
      runId?: string,
    ): Promise<void> {
      const name = tool.name ?? tool.id?.[tool.id.length - 1] ?? 'unknown';
      if (runId) toolNames.set(runId, name);
      if (options.onToolStart) await options.onToolStart(name, input);
    },

    async handleToolEnd(output: string, runId?: string): Promise<void> {
      const name = runId ? (toolNames.get(runId) ?? '') : '';
      if (runId) toolNames.delete(runId);
      if (options.onToolEnd) await options.onToolEnd(name, output);
    },

    async handleToolError(err: Error, runId?: string): Promise<void> {
      const name = runId ? (toolNames.get(runId) ?? '') : '';
      if (runId) toolNames.delete(runId);
      if (options.onToolError) await options.onToolError(name, err);
    },
  };
}
